river/model_selection/bandit.py

```diff
@@ -0,0 +1,120 @@
+"""
+
+This module contains model selection logic based on multi-armed bandits (MAB). The way the code is
+organised, the bandit logic is agnostic of the model selection aspect.
+
+"""
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from random import Random
+from typing import Iterator, List
+
+from river import base, metrics, utils
+
+from .base import ModelSelectionRegressor
+
+
+@dataclass
+class Arm:
+    """An arm in a multi-armed bandit.
+
+    It may also be referred to as a "lever".
+
+    """
+
+    index: int
+    metric: metrics.Metric
+    n_pulls: int = 0
+
+
+class Bandit:
+    """(Multi-armed) bandit (MAB) policy.
+
+    A bandit is composed of multiple arms. A policy is in charge of determining the best one.
+
+    """
+
+    def __init__(self, n_arms: int, metric: metrics.Metric):
+        self.arms = [Arm(index=i, metric=metric.clone()) for i in range(n_arms)]
+        self.metric = metric
+        self.best_arm = self.arms[0]
+        self.n_pulls = 0
+
+    def update(self, arm: Arm, **metric_kwargs):
+        self.n_pulls += 1
+        arm.n_pulls += 1
+        arm.metric.update(**metric_kwargs)
+        self.best_arm = max(
+            self.arms,
+            key=lambda arm: arm.metric.get()
+            if self.metric.bigger_is_better
+            else -arm.metric.get(),
+        )
+
+    @property
+    def ranking(self):
+        return [
+            arm.index
+            for arm in sorted(
+                self.arms,
+                key=lambda arm: arm.metric.get(),
+                reverse=self.metric.bigger_is_better,
+            )
+        ]
+
+    def __repr__(self):
+        return utils.pretty.print_table(
+            headers=["Ranking", self.metric.__class__.__name__, "Pulls", "Share",],
+            columns=[
+                [str(r) for r in self.ranking],
```

Comment:
I am not sure that you are really computing the ranking here. The ranking would be 
 `np.argsort(self.ranking)` (in numpy) or `sorted(range(len(r)), key=lambda k: r[k])` (in vanilla python). 
If you want to see an example of mismatch between the 2 you can try the docstring example with the following learning rates : 
`lrs = [0.0001, 0.001, 1e-05, 0.01]`

Reply:
Good catch! I'll add a unit test.